packages/store/src/patterns/patternMatchers.js

```diff
@@ -151,6 +154,107 @@ const makePatternKit = () => {
     // eslint-disable-next-line no-use-before-define
     HelpersByMatchTag[tag];
 
+  /**
+   * @typedef {string} Kind
+   * It is either a PassStyle other than 'tagged', or, if the underlying
+   * PassStyle is 'tagged', then the `getTag` value.
+   *
+   * We cannot further restrict this to only possible passStyles
+   * or known tags, because we wish to allow matching of tags that we
+   * don't know ahead of time. Do we need to separate the namespaces?
+   * TODO are we asking for trouble by lumping passStyles and tags
+   * together into kinds?
+   */
+
+  const kindMemo = new WeakMap();
+
+  /**
+   * Checks only recognized tags, and only if the tagged
+   * passes the invariants associated with that recognition.
+   *
+   * @param {Passable} tagged
+   * @param {Kind} tag
+   * @param {Checker} check
+   * @returns {boolean}
+   */
+  const checkTagged = (tagged, tag, check) => {
+    const matchHelper = maybeMatchHelper(tag);
+    if (matchHelper) {
+      // Buried here is the important case, where we process
+      // the various patternNodes
+      return matchHelper.checkIsWellFormed(tagged.payload, check);
+    }
+    switch (tag) {
+      case 'copySet': {
+        return checkCopySet(tagged, check);
+      }
+      case 'copyBag': {
+        return checkCopyBag(tagged, check);
+      }
+      case 'copyMap': {
+        return checkCopyMap(tagged, check);
+      }
+      default: {
+        return check(
+          false,
+          X`cannot check unrecognized tag ${q(tag)}: ${tagged}`,
+        );
+      }
+    }
+  };
+
+  /**
+   * Returns only a recognized kind, and only if the specimen passes the
+   * invariants associated with that recognition.
+   * Otherwise, `check(false, ...)` and returns undefined
+   *
+   * @param {Passable} specimen
+   * @param {Checker} [check]
+   * @returns {Kind | undefined}
+   */
+  const kindOf = (specimen, check = identChecker) => {
+    const passStyle = passStyleOf(specimen);
+    if (passStyle !== 'tagged') {
+      return passStyle;
+    }
+    // At this point we know that specimen is a well formed tagged record
```

Comment:
expanded comment:
```js
    // At this point we know that specimen is a well formed
    // as a tagged record, which is defined at the marshal level of abstraction,
    // since `passStyleOf` checks those invariants.
```